File: java-algorithm/src/test/java/com/example/algorithm/graph/BellmanFordTest.java
package com.example.algorithm.graph;

import lombok.extern.slf4j.Slf4j;
import org.junit.jupiter.api.AfterEach;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;
import org.mockito.junit.jupiter.MockitoExtension;

import static org.junit.jupiter.api.Assertions.assertEquals;

@ExtendWith(MockitoExtension.class)
@Slf4j
class BellmanFordTest {
    BellmanFord graph = new BellmanFord();
    BellmanFord.Vertex vertexS;
    BellmanFord.Vertex vertexT;
    BellmanFord.Vertex vertexX;
    BellmanFord.Vertex vertexY;
    BellmanFord.Vertex vertexZ;

    @BeforeEach
    void setUp() {
        vertexS = new BellmanFord.Vertex("S");
        vertexT = new BellmanFord.Vertex("T");
        vertexX = new BellmanFord.Vertex("X");
        vertexY = new BellmanFord.Vertex("Y");
        vertexZ = new BellmanFord.Vertex("Z");
        graph.addVertex(vertexS);
        graph.addVertex(vertexT);
        graph.addVertex(vertexX);
        graph.addVertex(vertexY);
        graph.addVertex(vertexZ);
        graph.addEdge(vertexS, vertexT, 6);
        graph.addEdge(vertexS, vertexY, 7);
        graph.addEdge(vertexT, vertexX, 5);
        graph.addEdge(vertexT, vertexY, 8);
        graph.addEdge(vertexT, vertexZ, -4);
        graph.addEdge(vertexX, vertexT, -2);
        graph.addEdge(vertexY, vertexX, -3);
        graph.addEdge(vertexY, vertexZ, 9);
        graph.addEdge(vertexZ, vertexS, 2);
        graph.addEdge(vertexZ, vertexX, 7);
    }

    @AfterEach
    void tearDown() {
    }

    @Test
    void testGetShortestPath() {
        graph.bellmanFordAlgorithm(vertexS);
        graph.printAllPaths();
        var pathT = graph.getShortestPath(vertexT);
        var pathX = graph.getShortestPath(vertexX);
        var pathY = graph.getShortestPath(vertexY);
        var pathZ = graph.getShortestPath(vertexZ);
        assertEquals(4, pathT.length);
        assertEquals(3, pathX.length);
        assertEquals(2, pathY.length);
        assertEquals(5, pathZ.length);
    }

    @Test
    void testGetShortestDistance() {
        graph.bellmanFordAlgorithm(vertexS);
        assertEquals(2, graph.getShortestDistance(vertexT));
        assertEquals(4, graph.getShortestDistance(vertexX));
        assertEquals(7, graph.getShortestDistance(vertexY));
        assertEquals(-2, graph.getShortestDistance(vertexZ));
    }
}

File: cpp-algorithm/src/graph/test/bellman_ford_test.cpp
#include "bellman_ford.h"

#include <gtest/gtest.h>

GTEST_TEST(BellmanFord, BellmanFord_Case1)
{
    auto vertex_s = BellmanFord::Vertex('S');
    auto vertex_t = BellmanFord::Vertex('T');
    auto vertex_x = BellmanFord::Vertex('X');
    auto vertex_y = BellmanFord::Vertex('Y');
    auto vertex_z = BellmanFord::Vertex('Z');

    const auto graph = new BellmanFord::Graph();

    graph->AddVertex(vertex_s);
    graph->AddVertex(vertex_t);
    graph->AddVertex(vertex_x);
    graph->AddVertex(vertex_y);
    graph->AddVertex(vertex_z);

    // S
    graph->AddEdge(vertex_s, vertex_t, 6);
    graph->AddEdge(vertex_s, vertex_y, 7);
    // T
    graph->AddEdge(vertex_t, vertex_x, 5);
    graph->AddEdge(vertex_t, vertex_y, 8);
    graph->AddEdge(vertex_t, vertex_z, -4);
    // X
    graph->AddEdge(vertex_x, vertex_t, -2);
    // Y
    graph->AddEdge(vertex_y, vertex_x, -3);
    graph->AddEdge(vertex_y, vertex_z, 9);
    // Z
    graph->AddEdge(vertex_z, vertex_s, 2);
    graph->AddEdge(vertex_z, vertex_x, 7);

    auto& source = vertex_s;
    source.distance = 0;
    graph->BellmanFordAlgorithm(source);

    // weight
    ASSERT_EQ(0, vertex_s.distance);
    ASSERT_EQ(2, vertex_t.distance);
    ASSERT_EQ(4, vertex_x.distance);
    ASSERT_EQ(7, vertex_y.distance);
    ASSERT_EQ(-2, vertex_z.distance);

    // predecessor
    ASSERT_EQ('X', vertex_t.predecessor->id);
    ASSERT_EQ('Y', vertex_x.predecessor->id);
    ASSERT_EQ('S', vertex_y.predecessor->id);
    ASSERT_EQ('T', vertex_z.predecessor->id);
}
